DEVELOPERS.md

```diff
@@ -0,0 +1,106 @@
+[Travis configuration](.travis.yml) for the codebase is also useful reference
+for the required tools and scripts, as it may be more up-to-date than this
+document.
+
+### Rebuilding Generated Code
+
+Some of the Trillian Go code is autogenerated from other files:
+
+ - [gRPC](http://www.grpc.io/) message structures are originally provided as
+   [protocol buffer](https://developers.google.com/protocol-buffers/) message
+   definitions.
+ - Some unit tests use mock implementations of interfaces; these are created
+   from the real implementations by [GoMock](https://github.com/golang/mock).
+ - Some enums have string-conversion methods (satisfying the `fmt.Stringer`
+   interface) created using the
+   [stringer](https://godoc.org/golang.org/x/tools/cmd/stringer) tool (`go get
+   golang.org/x/tools/cmd/stringer`).
+
+Re-generating mock or protobuffer files is only needed if you're changing
+the original files; if you do, you'll need to install the prerequisites:
+
+  - `mockgen` tool from https://github.com/golang/mock
+  - `stringer` tool from https://golang.org/x/tools/cmd/stringer
+  - `protoc`, [Go support for protoc](https://github.com/golang/protobuf),
+     [grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway) and
+     [protoc-gen-doc](https://github.com/pseudomuto/protoc-gen-doc).
+  - protocol buffer definitions for standard Google APIs:
+
+    ```bash
+    git clone https://github.com/googleapis/googleapis.git $GOPATH/src/github.com/googleapis/googleapis
+    ```
+
+and run the following:
+
+```bash
+go generate -x ./...  # hunts for //go:generate comments and runs them
+```
+
+### Updating Dependencies
+
+The Trillian codebase uses go.mod to declare fixed versions of its dependencies.
+With Go modules, updating a dependency simply involves running `go get`:
+```
+export GO111MODULE=on
+go get package/path       # Fetch the latest published version
+go get package/path@X.Y.Z # Fetch a specific published version
+go get package/path@HEAD  # Fetch the latest commit
+```
+
+To update ALL dependencies to the latest version run `go get -u`.
+Be warned however, that this may undo any selected versions that resolve issues in other non-module repos.
+
+While running `go build` and `go test`, go will add any ambiguous transitive dependencies to `go.mod`
+To clean these up run:
+```
+go mod tidy
+```
+
+### Running Codebase Checks
+
+The [`scripts/presubmit.sh`](scripts/presubmit.sh) script runs various tools
+and tests over the codebase.
+
+#### Install [golangci-lint](https://github.com/golangci/golangci-lint#local-installation).
+```bash
+go install github.com/golangci/golangci-lint/cmd/golangci-lint
+```
+
+#### Install [prototool](https://github.com/uber/prototool#installation)
+```bash
+go install github.com/uber/prototool/cmd/prototool
+```
+
+#### Run code generation, build, test and linters
+```bash
+./scripts/presubmit.sh
+```
+
+#### Or just run the linters alone
```

Comment:
Aren't there parameters for presubmit.sh to achieve this, rather than running individual tools directly (which seems like it could easily fall out-of-date)?